tools/src/main/scala/scala/scalanative/interflow/Combine.scala

```diff
@@ -0,0 +1,554 @@
+        combine(Ieq, ty, x, Val.Long(a - b))
+
+      // ((x xor a) == b) ==> (x == (a xor b))
+      case (Ieq, BinRef(Xor, x, Val.Char(a)), Val.Char(b)) =>
+        combine(Ieq, ty, x, Val.Char((a ^ b).toChar))
+      case (Ieq, BinRef(Xor, x, Val.Byte(a)), Val.Byte(b)) =>
+        combine(Ieq, ty, x, Val.Byte((a ^ b).toByte))
+      case (Ieq, BinRef(Xor, x, Val.Short(a)), Val.Short(b)) =>
+        combine(Ieq, ty, x, Val.Short((a ^ b).toShort))
+      case (Ieq, BinRef(Xor, x, Val.Int(a)), Val.Int(b)) =>
+        combine(Ieq, ty, x, Val.Int(a ^ b))
+      case (Ieq, BinRef(Xor, x, Val.Long(a)), Val.Long(b)) =>
+        combine(Ieq, ty, x, Val.Long(a ^ b))
+
+      // ((x xor true) == y) ==> (x != y)
+      case (Ieq, BinRef(Xor, x, Val.True), y) =>
+        combine(Ine, ty, x, y)
+
+      // (x == (y xor true)) ==> (x != y)
+      case (Ieq, x, BinRef(Xor, y, Val.True)) =>
+        combine(Ine, ty, x, y)
+
+      case (_, l, r) =>
+        delay(Op.Comp(comp, ty, r, l))
+    }
+  }
+
+  def combine(conv: Conv, ty: Type, value: Val)(implicit state: State): Val = {
+    import state.{materialize, delay, emit}
+
+    (conv, ty, value) match {
+      // trunc[iN] (trunc[iM] x) ==> trunc[iN] x if N < M
+      case (Trunc, Type.I(n, _), ConvRef(Trunc, Type.I(m, _), x)) if n < m =>
+        combine(Trunc, ty, x)
+
+      // sext[iN] (sext[iM] x) ==> sext[iN] x if N > M
+      case (Sext, Type.I(n, _), ConvRef(Sext, Type.I(m, _), x)) if n > m =>
+        combine(Sext, ty, x)
+
+      // zext[iN] (zext[iM] x) ==> zext[iN] x if N > M
+      case (Zext, Type.I(n, _), ConvRef(Zext, Type.I(m, _), x)) if n > m =>
+        combine(Zext, ty, x)
+
+      // ptrtoint[long] (inttoptr[long] x) ==> x
+      case (Ptrtoint, Type.Long, ConvRef(Inttoptr, Type.Long, x)) =>
+        x
+
+      // inttoptr[long] (ptrtoint[long] x) ==> x
+      case (Inttoptr, Type.Long, ConvRef(Ptrtoint, Type.Long, x)) =>
+        x
+
+      // bitcast[ty1] (bitcast(ty2) x) ==> bitcast[ty1] x
+      case (Bitcast, _, ConvRef(Bitcast, _, x)) =>
+        combine(Bitcast, ty, x)
+
+      case _ =>
+        delay(Op.Conv(conv, ty, value))
+    }
+  }
+
+  private def zero(ty: Type): Val =
+    Val.Zero(ty).canonicalize
+
+  private def minusOne(ty: Type): Val = ty match {
+    case Type.Byte   => Val.Byte(-1)
+    case Type.Short  => Val.Short(-1)
+    case Type.Int    => Val.Int(-1)
+    case Type.Long   => Val.Long(-1)
+    case Type.Float  => Val.Float(-1)
+    case Type.Double => Val.Double(-1)
+    case _           => unreachable
+  }
+
+  object PowerOf2 {
+    def unapply(v: Val): Option[Val] = {
+      v match {
+        case Val.Int(v) if isPowerOfTwo(v) =>
+          Some(Val.Int(jl.Integer.numberOfTrailingZeros(v)))
+        case Val.Long(v) if isPowerOfTwo(v) =>
+          Some(Val.Long(jl.Long.numberOfTrailingZeros(v)))
+        case _ =>
+          None
+      }
+    }
+
+    def isPowerOfTwo(x: Long): Boolean =
+      (x & (x - 1)) == 0
```

Comment:
That will return `true` for `Long.MinValue`, but `false` for any other negative Long value.

Reply:
Yep, it matches but the optimization that uses it still works even on min value. I renamed the helper method to be more explicit that it also matches min value.

Reply:
java.lang.Math contains methods that end with `Unsigned` but take normal longs and ints. With such convention the name of above function would be `isPowerOfTwoUnsigned`. Would that fit?

Reply:
Treating is an unsigned power of 2 makes perfect sense. Given that some of rules here work on signed integers as well, I'd probably keep the current name.